src/solver/tron.jl

```diff
@@ -0,0 +1,403 @@
+optimization:
+
+    min f(x)    s.t.    ℓ ≦ x ≦ u
+
+TRON is described in
+
+Chih-Jen Lin and Jorge J. Moré, *Newton's Method for Large Bound-Constrained
+Optimization Problems*, SIAM J. Optim., 9(4), 1100–1127, 1999.
+"""
+function tron(nlp :: AbstractNLPModel;
+              μ₀ :: Real=1e-2,
+              μ₁ :: Real=1.0,
+              σ :: Real=10.0,
+              verbose :: Bool=false,
+              itmax :: Int=10_000 * nlp.meta.nvar,
+              max_cgiter :: Int=nlp.meta.nvar,
+              cgtol :: Real=0.1,
+              fmin :: Real=-1e32,
+              timemax :: Real=60.0,
+              atol :: Real=1e-8,
+              rtol :: Real=1e-6,
+              fatol :: Real=0.0,
+              frtol :: Real=1e-12
+             )
+  ℓ = nlp.meta.lvar
+  u = nlp.meta.uvar
+  f(x) = obj(nlp, x)
+  g(x) = grad(nlp, x)
+  n = nlp.meta.nvar
+
+  iter = 0
+  start_time = time()
+  el_time = 0.0
+
+  # Preallocation
+  temp = zeros(n)
+  gpx = zeros(n)
+  xc = zeros(n)
+  Hs = zeros(n)
+
+  x = max.(ℓ, min.(nlp.meta.x0, u))
+  fx = f(x)
+  gx = g(x)
+  num_success_iters = 0
+
+  # Optimality measure
+  project_step!(gpx, x, -gx, ℓ, u)
+  πx = norm(gpx)
+  ϵ = atol + rtol * πx
+  optimal = πx <= ϵ
+  tired = iter >= itmax || el_time > timemax
+  unbounded = fx < fmin
+  stalled = false
+  status = ""
+
+  αC = 1.0
+  tr = TRONTrustRegion(min(max(1.0, 0.1 * norm(πx)), 100))
+  if verbose
+    @printf("%4s  %9s  %7s  %7s  %7s  %s\n", "Iter", "f", "π", "Radius", "Ratio", "CG-status")
+    @printf("%4d  %9.2e  %7.1e  %7.1e\n", iter, fx, πx, get_property(tr, :radius))
+  end
+  while !(optimal || tired || stalled)
+    # Current iteration
+    xc .= x
+    fc = fx
+    Δ = get_property(tr, :radius)
+
+    # Model
+    H = hess_op!(nlp, xc, temp)
+
+    # Cauchy starts here
+    αC, s = cauchy(x, H, gx, Δ, αC, ℓ, u, μ₀=μ₀, μ₁=μ₁, σ=σ)
+
+    s, Hs, cgits, cginfo = projected_newton!(x, H, gx, Δ, cgtol, s, ℓ, u, max_cgiter=max_cgiter)
+    slope = dot(gx,s)
+    qs = 0.5 * dot(s, Hs) + slope
+    fx = f(x)
+
+    try
+      ratio!(tr, nlp, fc, fx, qs, x, s, slope)
+    catch exc
+      status = exc.msg
+      stalled = true
+      continue
+    end
+
+    if num_success_iters == 0
+      tr.radius = min(Δ, norm(s))
+    end
+
+    # Update the trust region
+    update!(tr, norm(s))
+
+    if acceptable(tr)
+      num_success_iters += 1
+      gx = g(x)
+      project_step!(gpx, x, -gx, ℓ, u)
+      πx = norm(gpx)
+    end
+
+    # No post-iteration
+
+    if !acceptable(tr)
+      fx = fc
+      x .= xc
+    end
+
+    iter += 1
+    el_time = time() - start_time
+    tired = iter >= itmax ||  el_time > timemax
+    optimal = πx <= ϵ || fx == -Inf
+
+    verbose && @printf("%4d  %9.2e  %7.1e  %7.1e  %7.1e  %s\n", iter, fx, πx, Δ, Ared / Pred, cginfo)
+  end
+
+  status = if tired
+    iter >= itmax ? "maximum number of iterations" : "maximum elapsed time"
+  elseif πx <= ϵ
+    "first-order stationary"
+  elseif fx == -Inf
```

Comment:
This seems a bit dramatic. If `fx == -Inf`, won't there have been some kind of error earlier? How about a test of the form `fx ≤ -min(1, abs(f(x0))) / eps(eltype(x))`?